LibreHardwareMonitorLib/Hardware/RyzenSMU.cs

```diff
@@ -0,0 +1,625 @@
+                    return false;
+                }
+
+
+                // Step 2: Write zero (0) to the RSP register
+                WriteReg(_rsp_addr, 0);
+
+                // Step 3: Write the argument(s) into the argument register(s)
+                for (int i = 0; i < cmdArgs.Length; ++i)
+                    WriteReg(_args_addr + (uint)(i * 4), cmdArgs[i]);
+
+                // Step 4: Write the message Id into the Message ID register
+                WriteReg(_cmd_addr, msg);
+
+                // Step 5: Wait until the Response register is non-zero.
+                tmp = 0;
+                retries = SMU_RETRIES_MAX;
+                do
+                    if (! ReadReg(_rsp_addr, ref tmp))
+                    {
+                        _mutex.ReleaseMutex();
+                        return false;
+                    }
+                while (tmp == 0 && 0 != retries--);
+
+                if (retries == 0 && tmp != (uint)Status.OK)
+                {
+                    _mutex.ReleaseMutex();
+                    return false;
+                }
+
+                // Step 6: If the Response register contains OK, then SMU has finished processing  the message.
+
+                args = new uint[SMU_REQ_MAX_ARGS];
+                for (byte i = 0; i < SMU_REQ_MAX_ARGS; i++)
+                {
+                    if (! ReadReg(_args_addr + (uint)(i * 4), ref args[i]))
+                    {
+                        _mutex.ReleaseMutex();
+                        return false;
+                    }
+                }
+
+                ReadReg(_rsp_addr, ref tmp);
+                _mutex.ReleaseMutex();
+            }
+
+            return tmp == (uint)Status.OK;
+        }
+
+        private bool WriteReg(uint addr, uint data)
+        {
+            bool res = false;
+            if (Ring0.WaitPciBusMutex(10))
+            {
+                if (Ring0.WritePciConfig(0x00, SMU_PCI_ADDR_REG, addr) == true)
+                {
+                    res = Ring0.WritePciConfig(0x00, SMU_PCI_DATA_REG, data);
+                }
+            }
+            Ring0.ReleasePciBusMutex();
+            return res;
+        }
+
+        private bool ReadReg(uint addr, ref uint data)
+        {
+            bool res = false;
+            if (Ring0.WaitPciBusMutex(10))
+            {
+                if (Ring0.WritePciConfig(0x00, SMU_PCI_ADDR_REG, addr) == true)
+                {
+                    res = Ring0.ReadPciConfig(0x00, SMU_PCI_DATA_REG, out data);
+                }
+            }
+            Ring0.ReleasePciBusMutex();
+            return res;
+        }
+
+        public struct SmuSensorType
+        {
+            public string Name;
+            public SensorType Type;
+        }
+
+        private Dictionary<uint, Dictionary<uint, SmuSensorType>> _supported_pm_table_versions = new Dictionary<uint, Dictionary<uint, SmuSensorType>>()
+        {
+            {0x001E0004, new Dictionary<uint, SmuSensorType>()
```

Comment:
I think more will need to be added here, Zen 2 CPUs don't work with this.

Reply:
Yep, someone with a different CPU should just add their config to the Dictionary with the PM table version (eg. `0x001E0004`) the sensor index, name and type.

Reply:
You mentioned you used a table for these, any idea where to find others? It would be useful to get an idea on how to add these. Also to add support for some other CPUs.

Reply:
I think it is not made public by AMD.
It would be really useful for all other projects too.